data_structures/src/chain.rs

```diff
@@ -2902,6 +2911,112 @@ impl ReputationEngine {
     pub fn clear_threshold_cache(&self) {
         self.threshold_cache.borrow_mut().clear_threshold_cache()
     }
+
+    /// Get ARS keys ordered by reputation. If tie, order by pkh.
+    pub fn get_rep_ordered_ars_list(&self) -> Vec<PublicKeyHash> {
+        self.ars
+            .active_identities()
+            .cloned()
+            .sorted_by_key(|&pkh| std::cmp::Reverse((self.trs.get(&pkh), pkh)))
+            .collect()
+    }
+}
+
+#[allow(clippy::cast_possible_truncation, clippy::cast_sign_loss)]
+/// This function returns the result of: y = mx + K
+/// The result is rounded and low saturated in 0
+fn magic_line(x: f64, m: f64, k: f64) -> u32 {
+    let res = m * x + k;
+    if res < 0_f64 {
+        0
+    } else {
+        res.round() as u32
+    }
+}
+
+/// This function filter those identities which are reputed
+fn filter_reputed_identities(
+    identities: &[PublicKeyHash],
+    trs: &TotalReputationSet<PublicKeyHash, Reputation, Alpha>,
+) -> (Vec<PublicKeyHash>, u32) {
+    let mut total_rep = 0;
+    let filtered_identities = identities
+        .iter()
+        .filter_map(|pkh| {
+            // Identities with zero reputations will be excluded for the trapezoid calculations
+            let rep = trs.get(pkh).0;
+            if rep != 0 {
+                total_rep += rep;
+                Some(*pkh)
+            } else {
+                None
+            }
+        })
+        .collect();
+
+    (filtered_identities, total_rep)
+}
+
+#[allow(clippy::cast_precision_loss)]
+/// This function calculate the values and the total reputation
+/// for the upper triangle of the trapezoid
+fn calculate_trapezoid_triangle(m: f64, k: f64, identities_len: usize) -> (Vec<u32>, u32) {
+    let mut total_triangle_reputation = 0;
+    let mut triangle_reputation = vec![];
+
+    for i in 0..identities_len {
+        let calculated_rep = magic_line(i as f64, m, k);
+        total_triangle_reputation += calculated_rep;
+        triangle_reputation.push(calculated_rep);
+    }
+
+    (triangle_reputation, total_triangle_reputation)
+}
+
+#[allow(clippy::cast_possible_truncation, clippy::cast_precision_loss)]
+/// This functions calculate the conversion HashMap with the trapezoid distribution
```

Comment:
```suggestion
/// Calculate the conversion HashMap with the trapezoid distribution
```

Reply:
What's the conversion HashMap?

Reply:
A HashMap with the PublicKeyHashes and their new reputation (trapezoid reputation)